crates/subspace-service/src/sync_from_dsn.rs

```diff
@@ -40,38 +40,53 @@ const CHECK_ALMOST_SYNCED_INTERVAL: Duration = Duration::from_secs(1);
 /// Period of time during which node should be offline for DSN sync to kick-in
 const MIN_OFFLINE_PERIOD: Duration = Duration::from_secs(60);
 
-/// Trait representing a way to get pieces for DSN sync purposes
-#[async_trait]
-pub trait DsnSyncPieceGetter: fmt::Debug {
-    async fn get_piece(
-        &self,
-        piece_index: PieceIndex,
-    ) -> Result<Option<Piece>, Box<dyn Error + Send + Sync + 'static>>;
+/// Wrapper type for [`PieceProvider`], so it can implement [`PieceGetter`]
+pub struct DsnPieceGetter<PV: PieceValidator>(PieceProvider<PV>);
+
+impl<PV> fmt::Debug for DsnPieceGetter<PV>
+where
+    PV: PieceValidator,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("DsnPieceGetter")
+            .field(&format!("{:?}", self.0))
+            .finish()
+    }
 }
 
 #[async_trait]
-impl<T> DsnSyncPieceGetter for Arc<T>
+impl<PV> PieceGetter for DsnPieceGetter<PV>
 where
-    T: DsnSyncPieceGetter + Send + Sync + ?Sized,
+    PV: PieceValidator,
 {
-    async fn get_piece(
-        &self,
-        piece_index: PieceIndex,
-    ) -> Result<Option<Piece>, Box<dyn Error + Send + Sync + 'static>> {
-        self.as_ref().get_piece(piece_index).await
+    #[inline]
+    async fn get_piece(&self, piece_index: PieceIndex) -> anyhow::Result<Option<Piece>> {
+        Ok(self.0.get_piece_from_cache(piece_index).await)
+    }
+
+    #[inline]
+    async fn get_pieces<'a>(
+        &'a self,
+        piece_indices: Vec<PieceIndex>,
+    ) -> anyhow::Result<
+        Box<dyn Stream<Item = (PieceIndex, anyhow::Result<Option<Piece>>)> + Send + Unpin + 'a>,
+    > {
+        let stream = self
+            .0
+            .get_from_cache(piece_indices)
+            .await
+            .map(|(piece_index, maybe_piece)| (piece_index, Ok(maybe_piece)));
```

Comment:
Do we want to fall back to a random walk here (and in the method above)?

Reply:
Not on the node, this was also the behavior before this PR. Since we download the whole sector anyway, we should not need this